flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkBranch.java

```diff
@@ -0,0 +1,398 @@
+                    TableProperties.FORMAT_VERSION,
+                    String.valueOf(formatVersion)));
+
+    env =
+        StreamExecutionEnvironment.getExecutionEnvironment(
+                MiniClusterResource.DISABLE_CLASSLOADER_CHECK_CONFIG)
+            .enableCheckpointing(100);
+
+    tableLoader = catalogResource.tableLoader();
+  }
+
+  @Test
+  public void testWriteRow() throws Exception {
+    testWriteRow(null, DistributionMode.NONE);
+  }
+
+  @Test
+  public void testWriteRowWithTableSchema() throws Exception {
+    testWriteRow(SimpleDataUtil.FLINK_SCHEMA, DistributionMode.NONE);
+    verifyOtherBranchUnmodified();
+  }
+
+  private void testWriteRow(TableSchema tableSchema, DistributionMode distributionMode)
+      throws Exception {
+    List<Row> rows = createRows("");
+    DataStream<Row> dataStream = env.addSource(createBoundedSource(rows), ROW_TYPE_INFO);
+
+    FlinkSink.forRow(dataStream, SimpleDataUtil.FLINK_SCHEMA)
+        .table(table)
+        .tableLoader(tableLoader)
+        .tableSchema(tableSchema)
+        .toBranch(branch)
+        .distributionMode(distributionMode)
+        .append();
+
+    // Execute the program.
+    env.execute("Test Iceberg DataStream.");
+
+    SimpleDataUtil.assertTableRows(table, convertToRowData(rows), branch);
+    SimpleDataUtil.assertTableRows(
+        table,
+        ImmutableList.of(),
+        branch.equals(SnapshotRef.MAIN_BRANCH) ? "test-branch" : SnapshotRef.MAIN_BRANCH);
+
+    verifyOtherBranchUnmodified();
+  }
+
+  @Test
+  public void testCheckAndGetEqualityFieldIds() {
+    Assume.assumeTrue(formatVersion == 2);
+    table
+        .updateSchema()
+        .allowIncompatibleChanges()
+        .addRequiredColumn("type", Types.StringType.get())
+        .setIdentifierFields("type")
+        .commit();
+
+    DataStream<Row> dataStream =
+        env.addSource(new BoundedTestSource<>(ImmutableList.of()), ROW_TYPE_INFO);
+    FlinkSink.Builder builder =
+        FlinkSink.forRow(dataStream, SimpleDataUtil.FLINK_SCHEMA).table(table);
+
+    // Use schema identifier field IDs as equality field id list by default
+    Assert.assertEquals(
+        table.schema().identifierFieldIds(),
+        Sets.newHashSet(builder.checkAndGetEqualityFieldIds()));
+
+    // Use user-provided equality field column as equality field id list
+    builder.equalityFieldColumns(Lists.newArrayList("id"));
+    Assert.assertEquals(
+        Sets.newHashSet(table.schema().findField("id").fieldId()),
+        Sets.newHashSet(builder.checkAndGetEqualityFieldIds()));
+
+    builder.equalityFieldColumns(Lists.newArrayList("type"));
+    Assert.assertEquals(
+        Sets.newHashSet(table.schema().findField("type").fieldId()),
+        Sets.newHashSet(builder.checkAndGetEqualityFieldIds()));
+
+    verifyOtherBranchUnmodified();
+  }
+
+  @Test
+  public void testChangeLogOnIdKey() throws Exception {
+    Assume.assumeTrue(formatVersion == 2);
+    List<List<Row>> elementsPerCheckpoint =
```

Comment:
I still saw a lot of duplications with `TestFlinkIcebergSinkV2`. There are two differences
1) branch arg for `testChangeLogs`
2) verifyOtherBranchUnmodified

Might be easier if we have separate `TestFlinkIcebergSinkBase` and `TestFlinkIcebergSinkV2Base`. then we can have separate `TestFlinkIcebergSinkBranch` and `TestFlinkIcebergSinkV2Branch`.

Reply:
Good point, I did some more refactoring to further reduce the duplication. The common test cases with v2 will be shared (so the row data and the append will be the same) but we'll pass in a branch. For the non branch test cases we pass in just main, but parameterize based on all the existing properties (like file format, partitioning etc). For the branch ones we'll pass in just the branches with a fixed file format, partitioning. In this way we can reduce the duplication and get the benefit of not exponentially increasing the number of tests.